parser-typechecker/src/Unison/Runtime/Machine.hs

```diff
@@ -1289,7 +1289,7 @@ bprim2 !ustk !bstk IDXB i j = do
     Just x -> do
       poke ustk $ fromIntegral x
       ustk <- bump ustk
-      ustk <$ poke ustk 0
+      ustk <$ poke ustk 1
   pure (ustk, bstk)
```

Comment:
@dolio a few questions:

* Can you briefly explain `ustk` vs `bstk`?
* Is the final `poke ustk 0` or `poke ustk 1` the tag for the data type? Like `0` means `None` and `1` means `Some`?
* Is the idea that `Some 8` is represented as `[8, 1]` on the top of the `ustk`? Or should it be `[1,8]`? (Basically, should the tag be on the top of the stack) I'm a little confused because I looked at `IDXS` (`List.at`) and it seems to be done differently.
* Can you explain how it works to store things to the unboxed stack? 
  * Suppose I do `match Bytes.at 0 b with Some x -> x + 99`. How does that code know that the x will be found on `ustk`?
  * Suppose I do `match List.at 0 [1,2,3] with Some x -> x + 99`. How does that code know that the `x` value will be found on the `bstk`?

Reply:
The 'data' stack should be thought of as consisting of both `ustk` and `bstk`, although they are stored/indexed independently. `ustk` is where we store 64-bit unboxed stuff in a byte array, and `bstk` stores boxed stuff as an array of closures.

As you guessed, the final `poke ustk 1` is a sort of tag information indicating what is being returned. The builtin operation that makes use of this primop will test that to determine how to build the result. It's basically like an unboxed sum that the wrapper uses to build an actual sum value.

The ordering on the stack depends on how you're thinking about it. I have it set up so that they grow toward increasing array indices. So, the stack 'pointer' is a positive index into the relevant array, saying the highest index that has meaningful stack data. Offset positions in the stack are at lower indices, so position `i` on the stack is at index `p-i`, where `p` is the 'stack pointer' index. So as array indices, they are backwards, but the stack operations are set up so that you'd use the same numbering if the stack grew _down_, in which case things would be stored in the 'right' order.

None of your code examples do anything with the unboxed stack, except that pattern matching will use it for the implementation of testing the data type tag. `x` is not on the unboxed stack, it is a boxed integer. The builtins that implement addition will use the unboxed stack, but no value in surface-level unison is something on the unboxed stack, I think.

Reply:
> None of your code examples do anything with the unboxed stack, except that pattern matching will use it for the implementation of testing the data type tag. x is not on the unboxed stack, it is a boxed integer. 

@dolio hmm then can you explain this line, which seems to put the thing inside the `Just` onto the unboxed portion of the stack?

```
poke ustk $ fromIntegral x
```

Reply:
You're looking at the implementation of the machine instruction. But it is wrapped by code in `Unison.Runtime.Builtin` that implements the surface-level unison functions, which are not directly calling the machine instruction.

The builtin code reads the tag and value off the unboxed stack and builds an appropriate data value, which is what the corresponds to the surface level.

Reply:
Gotcha. So the machine instruction is technically returning an unboxed `Nat` with a tag of `0` or `1`, while the surface level function is calling that with some calling convention that boxes up the result appropriately.

Reply:
Yeah, that's right. There's always a wrapper around these instruction things. Some of them just wrap the instruction in something that looks more like the representation of compiled code. But many of them do this unpacking and packing stuff.

Reply:
Cool. Makes sense. Thanks for clarification.